server/gc/src/main/java/org/apache/accumulo/gc/GarbageCollectionAlgorithm.java

```diff
@@ -343,4 +357,82 @@ public void collect(GarbageCollectionEnvironment gce)
       deleteConfirmed(gce, candidateMap);
     }
   }
+
+  /**
+   * Track metadata rows read to help validate that gc scan has complete information to make a
+   * decision on deleting files
+   */
+  private static class MetadataReadTracker implements AutoCloseable {
+    private boolean hasDir = false;
+    private boolean hasPrevRow = false;
+    private Text row;
+
+    private boolean closed = false;
+
+    public MetadataReadTracker() {
+      this.row = null;
+    }
+
+    private void validate() {
+      Preconditions.checkState(hasDir && hasPrevRow,
+          "May not have fully read metadata for row, aborting this run. Validation results: %s",
+          this);
+    }
+
+    /**
+     * Initializes row tracking for the provided row. If a previous row was being tracked, it is
+     * checked that all expected metadata fields have been marked as seen. If all fields have not
+     * been marked seen, an IllegalStateException is thrown to halt further processing.
+     *
+     * @param candidate
+     *          check current row, initialize a new row to track
+     */
+    public void trackRow(final Text candidate) {
+      Preconditions.checkState(!closed);
+      Objects.requireNonNull(candidate);
+      if (row == null) {
+        row = candidate;
+      } else if (!row.equals(candidate)) {
+        validate();
+        hasPrevRow = false;
+        hasDir = false;
+        row = candidate;
+      }
+    }
+
+    /**
+     * Mark that the dir metadata entry seen for the current row being tracked.
+     */
+    public void markDirSeen() {
+      Preconditions.checkState(!closed);
+      hasDir = true;
+    }
+
+    /**
+     * Mark that the prevRow metadata entry seen for the current row being tracked.
+     */
+    public void markPrevRowSeen() {
+      Preconditions.checkState(!closed);
+      hasPrevRow = true;
+    }
+
+    /**
+     * Check that the final row processed is complete and then close tracker to additional
+     * processing.
+     */
+    @Override
+    public void close() {
+      Preconditions.checkState(!closed);
+      if (row != null) {
+        validate();
+      }
+      closed = true;
```

Comment:
It's nice to make close methods idempotent. If they are already closed, it doesn't need to be an error.

```suggestion
      if (!closed && row != null) {
        validate();
      }
      closed = true;
```